README.md
# ColorPicker

这是一个vscode拓展

使得魔兽争霸3的字符串前缀可以被颜色选择器识别和修改
后续也支持了一些常规的格式可以被颜色选择器识别和修改

支持以下语法

``` txt
    war3语法
         - |cffffffff
    color函数
         - color(255,255,255)
         - color(255,255,255,255)
         大写开头`Color`也可识别
    rgba函数
         - rgb(255,255,255)
         - rgba(255,255,255,0.5)
         大写也可以识别
    css
         - #ffffff
```

目前支持以下后缀格式的文件
|格式|war3语法|color函数|rgba函数|css|
|:---|---|---|---|---|
jass | 支持 | 支持 | - | - 
vjass | 支持 | 支持| - | - 
zinc | 支持 | 支持| - | - 
lua | 支持 | 支持 | - | 支持
ini | 支持 | - | - | 支持
fdf | 支持| - | - | - 
json | 支持| - | - | 支持
js | 支持 | 支持 | 支持 | 支持
ts | 支持 | 支持 | 支持 | 支持
xml | - | - | - | 支持
html | - | - | 支持 | 支持
css | - | - | 支持 | 支持

旧群被封 新群:
魔兽争霸3 lua作图交流群:1019770872


CHANGELOG.md
# Change Log

All notable changes to the "color4wc3text" extension will be documented in this file.

Check [Keep a Changelog](https://keepachangelog.com/) for recommendations on how to structure this file.

## [0.1.16]

- 修复了打开超大文件时,可能发生卡死的情况

## [0.1.14]

- 现在修改Color和RGBA不会影响大小写了

## [0.1.10]

- 支持了rgba和rgb

## [0.1.9]

- Color现在支持十六进制表述
- 修复js和ts支持
- 修复一直为黑色的bug

## [0.1.7]

- Color支持大写
- 对js和ts提供支持

## [0.1.1]

- color支持三个参数的场景

## [0.1.0]

- 修复color逗号的错误 添加css语法识别

## [0.0.10]

- 修复BUG并且更新依赖

## [0.0.8]

- 添加了fdf和json格式

## [0.0.7]

- 尝试修复可能和其他插件冲突的问题

## [0.0.6]

- 没啥变化 瞎鸡儿更新


src/main/document-color-provider.ts
/*
 * @Date: 2020-02-20 22:15:07
 * @LastEditTime: 2020-02-21 00:50:33
 */
/*
 * @Date: 2020-02-20 22:15:07
 * @LastEditTime: 2020-02-21 00:35:48
 */
import * as vscode from 'vscode';
import { languages } from './constant';

/// 颜色提供
const convertInt2Hex = (int: number) => {
  return Math.ceil(int * 255).toString(16).padStart(2, "0");
};
const color2JColorCode = (color: vscode.Color) => {
  if (color instanceof vscode.Color) {
    let r = color.red;
    let g = color.green;
    let b = color.blue;
    let a = color.alpha;
    let colorCodeString = convertInt2Hex(a) + convertInt2Hex(r) + convertInt2Hex(g) + convertInt2Hex(b);
    return colorCodeString;
  }
  return "00000000";
};
const color2ColorClassColorCode = (color: vscode.Color) => {
  if (color instanceof vscode.Color) {
    let r = color.red;
    let g = color.green;
    let b = color.blue;
    let a = color.alpha;
    let colorCodeString = ""+Math.ceil(r * 255)+","+Math.ceil(g * 255)+","+Math.ceil(b * 255)+","+Math.ceil(a * 255);
    return colorCodeString;
  }
  return "0,0,0,0";
};

class HexDocumentColorProvider implements vscode.DocumentColorProvider {

  /// 颜色改变到文档
  provideDocumentColors(document: vscode.TextDocument, token: vscode.CancellationToken): vscode.ProviderResult<vscode.ColorInformation[]> {
    let lineCount = document.lineCount;
    let colors = new Array<vscode.ColorInformation>();
    // new RegExp(/\|[cC][\da-fA-F]{8}.+?\|[rR]/, "g")
    let colorReg = new RegExp(/\|[cC][\da-fA-F]{8}/, "g");
    for (let i = 0; i < lineCount; i++) {
      let lineText = document.lineAt(i).text;
      let colotSet = lineText.match(colorReg);
      let posstion = 0;
      if (colotSet) {
        colotSet.forEach(x => {
          posstion = lineText.indexOf(x, posstion);
          let range = new vscode.Range(i, posstion, i, posstion + x.length);
          let a = Number.parseInt("0x" + lineText.substr(posstion + 2, 2)) / 255;
          let r = Number.parseInt("0x" + lineText.substr(posstion + 4, 2)) / 255;
          let g = Number.parseInt("0x" + lineText.substr(posstion + 6, 2)) / 255;
          let b = Number.parseInt("0x" + lineText.substr(posstion + 8, 2)) / 255;
          colors.push(new vscode.ColorInformation(range, new vscode.Color(r, g, b, a)));
          posstion += x.length;
        });
      }
    }
    return colors;
  }
  /// 文档改变到颜色
  provideColorPresentations(color: vscode.Color, context: { document: vscode.TextDocument; range: vscode.Range; }, token: vscode.CancellationToken): vscode.ProviderResult<vscode.ColorPresentation[]> {
    let r = color.red;
    let g = color.green;
    let b = color.blue;
    let a = color.alpha;
    let document = context.document;
    let range = context.range;
    let documentText = document.getText(range);
    return [new vscode.ColorPresentation(`${
      documentText.substr(0, 2)
      }${
      color2JColorCode(new vscode.Color(r, g, b, a))
      }${
      documentText.substring(10)
      }`)];

  }
}

class ColorDocumentColorProvider implements vscode.DocumentColorProvider {

  /// 颜色改变到文档
  provideDocumentColors(document: vscode.TextDocument, token: vscode.CancellationToken): vscode.ProviderResult<vscode.ColorInformation[]> {
    let lineCount = document.lineCount;
    let colors = new Array<vscode.ColorInformation>();
    let colorReg = new RegExp(/color\(((\d+)(,\d+){3})\)/, "g");
    for (let i = 0; i < lineCount; i++) {
      let lineText = document.lineAt(i).text;
      let colotSet = lineText.match(colorReg);
      let posstion = 0;
      if (colotSet) {
        colotSet.forEach(x => {
          let nums = x.match(new RegExp(/(\d+),(\d+),(\d+),(\d+)/));
          posstion = lineText.indexOf(x, posstion);
          let range = new vscode.Range(i, posstion, i, posstion + x.length);
          let r = Number.parseInt(nums?nums[1]:"255") / 255;
          let g = Number.parseInt(nums?nums[2]:"255") / 255;
          let b = Number.parseInt(nums?nums[3]:"255") / 255;
          let a = Number.parseInt(nums?nums[4]:"255") / 255;
          colors.push(new vscode.ColorInformation(range, new vscode.Color(r, g, b, a)));
          posstion += x.length;
        });
      }
    }
    return colors;
  }
  /// 文档改变到颜色
  provideColorPresentations(color: vscode.Color, context: { document: vscode.TextDocument; range: vscode.Range; }, token: vscode.CancellationToken): vscode.ProviderResult<vscode.ColorPresentation[]> {
    let r = color.red;
    let g = color.green;
    let b = color.blue;
    let a = color.alpha;
    let document = context.document;
    let range = context.range;
    return [new vscode.ColorPresentation(`color(${
      color2ColorClassColorCode(new vscode.Color(r, g, b, a))
      })`)];

  }
}
languages.forEach(language=>{
  // if (language !== "jass" && language !== "wurst"){
    vscode.languages.registerColorProvider(language, new HexDocumentColorProvider);
  // }
  if (language !== "wurst" && language !== "ini")
  {
    vscode.languages.registerColorProvider(language, new ColorDocumentColorProvider);
  }
});


src/main/color-provider.ts
import * as vscode from 'vscode';


/// 颜色提供
const convertInt2Hex = (int: number) => {
  return Math.ceil(int * 255).toString(16).padStart(2, "0");
};
const color2CssColorCode = (color: vscode.Color) => {
  if (color instanceof vscode.Color) {
    let r = color.red;
    let g = color.green;
    let b = color.blue;
    let colorCodeString =  convertInt2Hex(r) + convertInt2Hex(g) + convertInt2Hex(b);
    return colorCodeString;
  }
  return "000000";
};
const color2JColorCode = (color: vscode.Color) => {
  if (color instanceof vscode.Color) {
    let r = color.red;
    let g = color.green;
    let b = color.blue;
    let a = color.alpha;
    let colorCodeString = convertInt2Hex(a) + convertInt2Hex(r) + convertInt2Hex(g) + convertInt2Hex(b);
    return colorCodeString;
  }
  return "00000000";
};
const ceilToStr = (int: number,isHex?: boolean) => {
  if (isHex) {
    return '0x'+Math.ceil(int * 255).toString(16).padStart(2, "0");
  }
  return Math.ceil(int * 255).toString(10);
}

const color2ColorClassColorCode = (color: vscode.Color,flags?:boolean[]) => {
  if (color instanceof vscode.Color) {
    let r = color.red;
    let g = color.green;
    let b = color.blue;
    let a = color.alpha;
    let colorCodeString:string
    if (flags) {
      colorCodeString = ceilToStr(r,flags[0])+", "+ceilToStr(g,flags[1])+", "+ceilToStr(b,flags[2])+", "+ceilToStr(a,flags[3]);
    }else{
      colorCodeString = ceilToStr(r)+", "+ceilToStr(g)+", "+ceilToStr(b)+", "+ceilToStr(a);
    }
    return colorCodeString;
  }
  return "0, 0, 0, 255";
};
const color2Color3ClassColorCode = (color: vscode.Color,flags?:boolean[]) => {
  if (color instanceof vscode.Color) {
    let r = color.red;
    let g = color.green;
    let b = color.blue;
    let colorCodeString:string
    if (flags) {
      colorCodeString = ceilToStr(r,flags[0])+", "+ceilToStr(g,flags[1])+", "+ceilToStr(b,flags[2]);
    }else{
      colorCodeString = ceilToStr(r)+", "+ceilToStr(g)+", "+ceilToStr(b);
    }
    return colorCodeString;
  }
  return "0, 0, 0";
};

const numberFormat = (str: string,base?: number) => {
  if (str.startsWith("0x") || str.startsWith("0X")) {
    return Number.parseInt(str)
  }
  if (base) {
    if (base == 16) {
      return Number.parseInt("0x" +str)
    }else{
      Number.parseInt(str)
    }
  }else{
    if (Number.parseInt(str).toString()==str) {
      return Number.parseInt(str)
    }
    if (str.match(new RegExp(/\[0-9a-fA-F]{2}/, "g"))) {
      return Number.parseInt("0x" +str)
    }
  }
  return Number.parseInt(str)
};

const getFuncName = (document: vscode.TextDocument, range: vscode.Range) => {
  let name = document.getText(range);
  return name.substring(0, name.indexOf('(')) 
}

class HexDocumentColorProvider implements vscode.DocumentColorProvider {
  /// 颜色改变到文档
  provideDocumentColors(document: vscode.TextDocument, token: vscode.CancellationToken): vscode.ProviderResult<vscode.ColorInformation[]> {
    let lineCount = document.lineCount;
    let colors = new Array<vscode.ColorInformation>();
    let colorReg = new RegExp(/\|[cC][0-9a-fA-F]{8}/, "g");
    for (let i = 0; i < lineCount; i++) {
      let lineText = document.lineAt(i).text;
      let colotSet = lineText.match(colorReg);
      let posstion = 0;
      if (colotSet) {
        colotSet.forEach(x => {
          posstion = lineText.indexOf(x, posstion);
          let range = new vscode.Range(i, posstion, i, posstion + x.length);
          let a = numberFormat( lineText.substring(posstion + 2, posstion + 2 + 2),16) / 255;
          let r = numberFormat( lineText.substring(posstion + 4, posstion + 4 + 2),16) / 255;
          let g = numberFormat( lineText.substring(posstion + 6, posstion + 6 + 2),16) / 255;
          let b = numberFormat( lineText.substring(posstion + 8, posstion + 8 + 2),16) / 255;
          colors.push(new vscode.ColorInformation(range, new vscode.Color(r, g, b, a)));
          posstion += x.length;
        });
      }
    }
    return colors;
  }
  /// 文档改变到颜色
  provideColorPresentations(color: vscode.Color, context: { document: vscode.TextDocument; range: vscode.Range; }, token: vscode.CancellationToken): vscode.ProviderResult<vscode.ColorPresentation[]> {
    let r = color.red;
    let g = color.green;
    let b = color.blue;
    let a = color.alpha;
    let document = context.document;
    let range = context.range;
    let documentText = document.getText(range);
    return [new vscode.ColorPresentation(`${
      documentText.substring(0, 2)
      }${
      color2JColorCode(new vscode.Color(r, g, b, a))
      }${
      documentText.substring(10)
      }`)];

  }
}


class CssDocumentColorProvider implements vscode.DocumentColorProvider {
  /// 颜色改变到文档
  provideDocumentColors(document: vscode.TextDocument, token: vscode.CancellationToken): vscode.ProviderResult<vscode.ColorInformation[]> {
    let lineCount = document.lineCount;
    let colors = new Array<vscode.ColorInformation>();
    let colorReg = new RegExp(/#[\da-fA-F]{6}/, "g");
    for (let i = 0; i < lineCount; i++) {
      let lineText = document.lineAt(i).text;
      let colotSet = lineText.match(colorReg);
      let posstion = 0;
      if (colotSet) {
        colotSet.forEach(x => {
          posstion = lineText.indexOf(x, posstion);
          let range = new vscode.Range(i, posstion, i, posstion + x.length);
          let a = 255 / 255;
          let r = numberFormat( lineText.substring(posstion + 1, posstion + 1 + 2),16) / 255;
          let g = numberFormat( lineText.substring(posstion + 3, posstion + 3 + 2),16) / 255;
          let b = numberFormat( lineText.substring(posstion + 5, posstion + 5 + 2),16) / 255;
          colors.push(new vscode.ColorInformation(range, new vscode.Color(r, g, b, a)));
          posstion += x.length;
        });
      }
    }
    return colors;
  }
  /// 文档改变到颜色
  provideColorPresentations(color: vscode.Color, context: { document: vscode.TextDocument; range: vscode.Range; }, token: vscode.CancellationToken): vscode.ProviderResult<vscode.ColorPresentation[]> {
    let r = color.red;
    let g = color.green;
    let b = color.blue;
    let a = color.alpha;
    return [new vscode.ColorPresentation(`#${
      color2CssColorCode(new vscode.Color(r, g, b, a))
    }`)]
  }
}


class ColorDocumentColorProvider implements vscode.DocumentColorProvider {
  /// 颜色改变到文档
  provideDocumentColors(document: vscode.TextDocument, token: vscode.CancellationToken): vscode.ProviderResult<vscode.ColorInformation[]> {
    let lineCount = document.lineCount;
    let colors = new Array<vscode.ColorInformation>();
    let colorReg = new RegExp(/[cC]olor\( *\w+( *, *\w+ *){3}\)/, "g");
    for (let i = 0; i < lineCount; i++) {
      let lineText = document.lineAt(i).text;
      let colotSet = lineText.match(colorReg);
      let posstion = 0;
      if (colotSet) {
        colotSet.forEach(x => {
          let nums = x.match(new RegExp(/ *(\w+) *, *(\w+) *, *(\w+) *, *(\w+) */));
          if (nums) {
            posstion = lineText.indexOf(x, posstion);
            let range = new vscode.Range(i, posstion, i, posstion + x.length);
            let r = numberFormat(nums[1]) / 255;
            let g = numberFormat(nums[2]) / 255;
            let b = numberFormat(nums[3]) / 255;
            let a = numberFormat(nums[4]) / 255;
            
            colors.push(new vscode.ColorInformation(range, new vscode.Color(r, g, b, a)));
            posstion += x.length;
          }
        });
      }
    }
    return colors;
  }
  /// 文档改变到颜色
  provideColorPresentations(color: vscode.Color, context: { document: vscode.TextDocument; range: vscode.Range; }, token: vscode.CancellationToken): vscode.ProviderResult<vscode.ColorPresentation[]> {
    let r = color.red;
    let g = color.green;
    let b = color.blue;
    let a = color.alpha;
    let name = getFuncName(context.document, context.range);
    return [new vscode.ColorPresentation(name+`(${
        color2ColorClassColorCode(new vscode.Color(r, g, b, a))
      })`)];

  }
}

class Color3DocumentColorProvider implements vscode.DocumentColorProvider {
  /// 颜色改变到文档
  provideDocumentColors(document: vscode.TextDocument, token: vscode.CancellationToken): vscode.ProviderResult<vscode.ColorInformation[]> {
    let lineCount = document.lineCount;
    let colors = new Array<vscode.ColorInformation>();
    let colorReg = new RegExp(/[cC]olor\( *[0-9A-Fa-fxX]+( *, *[0-9A-Fa-fxX]+ *){2}\)/, "g");
    for (let i = 0; i < lineCount; i++) {
      let lineText = document.lineAt(i).text;
      let colotSet = lineText.match(colorReg);
      let posstion = 0;
      if (colotSet) {
        colotSet.forEach(x => {
          let nums = x.match(new RegExp(/ *([0-9A-Fa-fxX]+) *, *([0-9A-Fa-fxX]+) *, *([0-9A-Fa-fxX]+) */));
          if(nums)
          {
            posstion = lineText.indexOf(x, posstion);
            let range = new vscode.Range(i, posstion, i, posstion + x.length);
            let r = numberFormat(nums[1]) / 255;
            let g = numberFormat(nums[2]) / 255;
            let b = numberFormat(nums[3]) / 255;
            let a = 1;
            colors.push(new vscode.ColorInformation(range, new vscode.Color(r, g, b, a)));
            posstion += x.length;
          }
        });
      }
    }
    return colors;
  }
  /// 文档改变到颜色
  provideColorPresentations(color: vscode.Color, context: { document: vscode.TextDocument; range: vscode.Range; }, token: vscode.CancellationToken): vscode.ProviderResult<vscode.ColorPresentation[]> {
    let r = color.red;
    let g = color.green;
    let b = color.blue;
    let a = color.alpha;
    let name = getFuncName(context.document, context.range);
    return [new vscode.ColorPresentation(name + `(${
        color2Color3ClassColorCode(new vscode.Color(r, g, b, a))
      })`)];
  }
}

class RGBADocumentColorProvider implements vscode.DocumentColorProvider {
  /// 颜色改变到文档
  provideDocumentColors(document: vscode.TextDocument, token: vscode.CancellationToken): vscode.ProviderResult<vscode.ColorInformation[]> {
    let lineCount = document.lineCount;
    let colors = new Array<vscode.ColorInformation>();
    let colorReg = new RegExp(/[rR][gG][bB][aA]\( *\w+( *, *\w+ *){2} *, *\d*\.?\d* *\)/, "g");
    for (let i = 0; i < lineCount; i++) {
      let lineText = document.lineAt(i).text;
      let colotSet = lineText.match(colorReg);
      let posstion = 0;
      if (colotSet) {
        colotSet.forEach(x => {
          let nums = x.match(new RegExp(/ *(\w+) *, *(\w+) *, *(\w+) *, *(\d*\.?\d*) */));
          if (nums) {
            posstion = lineText.indexOf(x, posstion);
            let range = new vscode.Range(i, posstion, i, posstion + x.length);
            let r = numberFormat(nums[1]) / 255;
            let g = numberFormat(nums[2]) / 255;
            let b = numberFormat(nums[3]) / 255;
            let a = Number.parseFloat(nums[4]) ;
            
            colors.push(new vscode.ColorInformation(range, new vscode.Color(r, g, b, a)));
            posstion += x.length;
          }
        });
      }
    }
    return colors;
  }
  /// 文档改变到颜色
  provideColorPresentations(color: vscode.Color, context: { document: vscode.TextDocument; range: vscode.Range; }, token: vscode.CancellationToken): vscode.ProviderResult<vscode.ColorPresentation[]> {
    let r = color.red;
    let g = color.green;
    let b = color.blue;
    let a = color.alpha;
    let name = getFuncName(context.document, context.range);
    return [new vscode.ColorPresentation(name+`(${
      color2Color3ClassColorCode(new vscode.Color(r, g, b, 1)) + ', ' + a.toFixed(2)
      })`)];

  }
}


class RGBDocumentColorProvider implements vscode.DocumentColorProvider {
  /// 颜色改变到文档
  provideDocumentColors(document: vscode.TextDocument, token: vscode.CancellationToken): vscode.ProviderResult<vscode.ColorInformation[]> {
    let lineCount = document.lineCount;
    let colors = new Array<vscode.ColorInformation>();
    let colorReg = new RegExp(/[rR][gG][bB]\( *\w+( *, *\w+ *){2}\)/, "g");
    for (let i = 0; i < lineCount; i++) {
      let lineText = document.lineAt(i).text;
      let colotSet = lineText.match(colorReg);
      let posstion = 0;
      if (colotSet) {
        colotSet.forEach(x => {
          let nums = x.match(new RegExp(/ *(\w+) *, *(\w+) *, *(\w+) */));
          if (nums) {
            posstion = lineText.indexOf(x, posstion);
            let range = new vscode.Range(i, posstion, i, posstion + x.length);
            let r = numberFormat(nums[1]) / 255;
            let g = numberFormat(nums[2]) / 255;
            let b = numberFormat(nums[3]) / 255;
            
            colors.push(new vscode.ColorInformation(range, new vscode.Color(r, g, b, 1)));
            posstion += x.length;
          }
        });
      }
    }
    return colors;
  }
  /// 文档改变到颜色
  provideColorPresentations(color: vscode.Color, context: { document: vscode.TextDocument; range: vscode.Range; }, token: vscode.CancellationToken): vscode.ProviderResult<vscode.ColorPresentation[]> {
    let r = color.red;
    let g = color.green;
    let b = color.blue;
    let a = color.alpha;
    let name = getFuncName(context.document, context.range);
    return [new vscode.ColorPresentation(name+`(${
      color2Color3ClassColorCode(new vscode.Color(r, g, b, 1))
      })`)];

  }
}

const hexSupportLanguages = ["jass","lua","ini","vjass","zinc","fdf","json",'js',"javascript","typescript"];
const colorSupportLanguages = ["jass","lua","vjass","zinc",'js',"javascript","typescript"];
const cssSupportLanguages = ["css","html","xml","json",'js',"javascript","typescript","lua","ini"];
const rgbaSupportLanguages = ["css","html",'js',"javascript","typescript"];

hexSupportLanguages.forEach(language=>{
  vscode.languages.registerColorProvider(language, new HexDocumentColorProvider);
});
colorSupportLanguages.forEach(language=>{
  vscode.languages.registerColorProvider(language, new ColorDocumentColorProvider);
  vscode.languages.registerColorProvider(language, new Color3DocumentColorProvider);
});
cssSupportLanguages.forEach(language=>{
  vscode.languages.registerColorProvider(language, new CssDocumentColorProvider);
});
rgbaSupportLanguages.forEach(language=>{
  vscode.languages.registerColorProvider(language, new RGBADocumentColorProvider);
  vscode.languages.registerColorProvider(language, new RGBDocumentColorProvider);
});
